pygfx/utils/transform.py

```diff
@@ -0,0 +1,212 @@
+
+    """
+
+    def __init__(self, compute_fn=None) -> None:
+        self.compute_fn = compute_fn
+        self.last_updated = None
+        self.cached_value = None
+
+    def __get__(self, obj: "AffineBase", objtype=None):
+        if obj is None:
+            return self
+
+        if self.last_updated is None or obj.last_modified > self.last_updated:
+            self.cached_value = self.compute_fn(obj)
+            self.last_updated = obj.last_modified
+
+        return self.cached_value
+
+    def __call__(self, obj):
+        """Helper to support wrapping by @property."""
+        return self.__get__(obj)
+
+
+class AffineBase:
+    last_modified: int
+    matrix: np.ndarray
+
+    @property
+    @cached
+    def inverse_matrix(self):
+        return np.linalg.inv(self.matrix)
+    
+    @cached
+    def _decomposed(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        return pla.matrix_decompose(self.matrix)
+
+    @property
+    def position(self) -> np.ndarray:
+        return self._decomposed[0]
+
+    @property
+    def rotation(self) -> np.ndarray:
+        return self._decomposed[1]
+
+    @property
+    def scale(self) -> np.ndarray:
+        return self._decomposed[2]
+
+    def __array__(self, dtype=None):
+        return self.matrix.astype(dtype)
+
+
+
+class AffineTransform(AffineBase):
+    """An affine tranformation"""
+
+    def __init__(
+        self, matrix=None, /, *, position=None, rotation=None, scale=None
+    ) -> None:
+        super().__init__()
+        self.last_modified = perf_counter_ns()
+
+        if matrix is None:
+            matrix = np.eye(4, dtype=float)
+
+        self.matrix = np.asarray(matrix)
+
+        if position is not None:
+            self.position = position
+
+        if rotation is not None:
+            self.rotation = rotation
+
+        if scale is not None:
+            self.scale = scale
+
+    def flag_update(self):
+        self.last_modified = perf_counter_ns()
+
+    @AffineBase.position.setter
+    def position(self, value):
+        self.matrix[:-1, -1] = value
+        self.last_modified = perf_counter_ns()
+
+    @AffineBase.rotation.setter
+    def rotation(self, value):
+        position = self.position
+        scale = self.scale
+
+        pla.matrix_make_transform(position, value, scale, out=self.matrix)
+        self.last_modified = perf_counter_ns()
+
+    @AffineBase.scale.setter
+    def scale(self, value):
+        position = self.position
+        rotation = self.rotation
+
+        pla.matrix_make_transform(position, rotation, value, out=self.matrix)
+        self.last_modified = perf_counter_ns()
+
+    def __matmul__(self, other):
+        if isinstance(other, AffineTransform):
+            return AffineTransform(self.matrix @ other.matrix)
+
+        return np.asarray(self) @ other
+
+    def look_at(self, target) -> None:
+        rotation = pla.matrix_make_look_at(self.position, target, (0, 1, 0))
```

Comment:
I removed `WorldObject.up` because it seemed odd xD. This is another indicator that `AffineTransform.look_at` should really be part of `WorldObject` and not `AffineTransform`. The transform currently has no notion of what `up` is. Instead, this is contextual info that exists on the `WorldObject` (depending on its type).

It's also a quite isolated/different method from all the others, so I would probably not make it a dedicated function in the first place. Instead I would just make it part of the logic where it is used.

Reply:
Yeah, that makes sense 👍   We took the `WorldObject.up` property from ThreeJS. Technically it could also be an arg for `look_at()`, but making it a property allows you to set it once, and then the rest just works.